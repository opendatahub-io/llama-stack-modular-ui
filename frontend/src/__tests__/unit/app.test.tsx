import * as React from 'react';
import '@testing-library/jest-dom';
import { act, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from '@app/index';

// Mock the auth service
jest.mock('../../app/services/authService.ts', () => ({
  authService: {
    isAuthenticated: jest.fn(() => false),
    handleAuthenticationCheck: jest.fn(() => Promise.resolve(true)), // Mock OAuth as disabled (immediate access)
    loadConfig: jest.fn(() => Promise.resolve({ oauthEnabled: false })),
    getToken: jest.fn(() => null),
    clearToken: jest.fn(),
    addAuthListener: jest.fn(() => jest.fn()), // Return a mock unsubscribe function
  },
}));

describe('App tests', () => {
  beforeAll(() => {
    global.XMLHttpRequest = jest.fn(() => ({
      open: jest.fn(),
      send: jest.fn(),
      setRequestHeader: jest.fn(),
      abort: jest.fn(),
      readyState: 4,
      status: 200,
      responseText: JSON.stringify({}),
      onreadystatechange: null,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    })) as any;

    window.HTMLElement.prototype.scrollIntoView = jest.fn();
  });

  test('should render default App component', async () => {
    let component;

    await act(async () => {
      component = render(<App />);
    });

    // Wait for async operations to complete
    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'Global navigation' })).toBeInTheDocument();
    });

    expect(component.asFragment()).toMatchSnapshot();
  }, 7000);

  it('should render a nav-toggle button', async () => {
    await act(async () => {
      render(<App />);
    });

    // Wait for async operations to complete
    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'Global navigation' })).toBeVisible();
    });
  });

  // I'm fairly sure that this test not going to work properly no matter what we do since JSDOM doesn't actually
  // draw anything. We could potentially make something work, likely using a different test environment, but
  // using Cypress for this kind of test would be more efficient.
  it.skip('should hide the sidebar on smaller viewports', () => {
    Object.defineProperty(window, 'innerWidth', { writable: true, configurable: true, value: 600 });

    render(<App />);

    window.dispatchEvent(new Event('resize'));

    expect(screen.queryByRole('link', { name: 'Chatbot' })).not.toBeInTheDocument();
  });

  it('should expand the sidebar on larger viewports', async () => {
    await act(async () => {
      render(<App />);
    });

    // Wait for async operations to complete first
    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'Global navigation' })).toBeInTheDocument();
    });

    await act(async () => {
      window.dispatchEvent(new Event('resize'));
      // eslint-disable-next-line no-promise-executor-return
      await new Promise((r) => setTimeout(r, 0)); // ensure full state flush
    });

    expect(screen.getByRole('link', { name: 'Chatbot' })).toBeVisible();
  });

  it('should hide the sidebar when clicking the nav-toggle button', async () => {
    const user = userEvent.setup();

    await act(async () => {
      render(<App />);
    });

    // Wait for async operations to complete first
    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'Global navigation' })).toBeInTheDocument();
    });

    await act(async () => {
      window.dispatchEvent(new Event('resize'));
      // eslint-disable-next-line no-promise-executor-return
      await new Promise((resolve) => setTimeout(resolve, 0)); // allow React to process layout updates
    });

    const button = screen.getByRole('button', { name: 'Global navigation' });

    expect(screen.getByRole('link', { name: 'Chatbot' })).toBeVisible();

    await user.click(button);

    expect(screen.queryByRole('link', { name: 'Chatbot' })).not.toBeInTheDocument();
  });
});
