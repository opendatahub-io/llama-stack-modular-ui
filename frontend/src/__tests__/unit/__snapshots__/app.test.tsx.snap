// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`App tests should render default App component 1`] = `
<DocumentFragment>
  <div
    class="pf-v6-c-page"
  >
    <div
      class="pf-v6-c-skip-to-content"
    >
      <a
        class="pf-v6-c-button pf-m-primary"
        data-ouia-component-id="OUIA-Generated-Button-primary-1"
        data-ouia-component-type="PF6/Button"
        data-ouia-safe="true"
        href="#primary-app-container"
      >
        <span
          class="pf-v6-c-button__text"
        >
          Skip to Content
        </span>
      </a>
    </div>
    <header
      class="pf-v6-c-masthead pf-m-display-inline-on-md"
    >
      <div
        class="pf-v6-c-masthead__main"
      >
        <span
          class="pf-v6-c-masthead__toggle"
        >
          <button
            aria-expanded="true"
            aria-label="Global navigation"
            class="pf-v6-c-button pf-m-plain"
            data-ouia-component-id="OUIA-Generated-Button-plain-1"
            data-ouia-component-type="PF6/Button"
            data-ouia-safe="true"
            type="button"
          >
            <span
              class="pf-v6-c-button__icon"
            >
              <svg
                aria-hidden="true"
                class="pf-v6-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 448 512"
                width="1em"
              >
                <path
                  d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
                />
              </svg>
            </span>
          </button>
        </span>
        <div
          class="pf-v6-c-masthead__brand"
          data-codemods="true"
        >
          <span
            class="pf-v6-c-masthead__logo"
            data-codemods="true"
          >
            <img
              alt="Logo"
              src="test-file-stub"
            />
          </span>
        </div>
      </div>
    </header>
    <div
      aria-hidden="false"
      class="pf-v6-c-page__sidebar pf-m-expanded"
      id="page-sidebar"
    >
      <div
        class="pf-v6-c-page__sidebar-body"
      >
        <nav
          aria-label="Global"
          class="pf-v6-c-nav"
          data-ouia-component-id="OUIA-Generated-Nav-1"
          data-ouia-component-type="PF6/Nav"
          data-ouia-safe="true"
          id="nav-primary-simple"
        >
          <ul
            class="pf-v6-c-nav__list"
            id="nav-list-simple"
            role="list"
          >
            <li
              class="pf-v6-c-nav__item"
              data-ouia-component-id="OUIA-Generated-NavItem-1"
              data-ouia-component-type="PF6/NavItem"
              data-ouia-safe="true"
            >
              <a
                aria-current="page"
                class="pf-v6-c-nav__link pf-m-current active"
                data-discover="true"
                href="/"
              >
                Chatbot
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
    <div
      class="pf-v6-c-page__main-container pf-m-fill"
    >
      <main
        class="pf-v6-c-page__main"
        id="primary-app-container"
        tabindex="-1"
      >
        <svg
          aria-label="Contents"
          aria-valuetext="Loading..."
          class="pf-v6-c-spinner pf-m-sm"
          role="progressbar"
          viewBox="0 0 100 100"
        >
          <circle
            class="pf-v6-c-spinner__path"
            cx="50"
            cy="50"
            fill="none"
            r="45"
          />
        </svg>
      </main>
    </div>
  </div>
</DocumentFragment>
`;
